### PR TITLE
Functions: Development: Run configuration for Azure Functions host #161

### PR DESCRIPTION
**Pull request by maartenba**
Adds a run/debug configurations for Azure Functions as described in #78 / #161 

Comes with:
* Settings page for specifying Azure Functions core tools path
* Run/debug configuration
* Provider for context menu in solution explorer

Demo

### CODE REVIEW

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/ui/AzureFunctionsConfigurationPanel.kt [hunk 1]
```diff
@@ -0,0 +1,80 @@
+/**
+ * Copyright (c) 2018 JetBrains s.r.o.
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/com/microsoft/intellij/configuration/ui/AzureFunctionsConfigurationPanel.kt]: Year: 2019 - here and below in all new files :)

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/GitHubReleases.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+import com.google.gson.annotations.SerializedName
+import retrofit2.Call
+import retrofit2.Retrofit
+import retrofit2.converter.gson.GsonConverterFactory
+import retrofit2.http.GET
+import retrofit2.http.Headers
+import retrofit2.http.Url
+
+interface GitHubReleasesService {
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/GitHubReleases.kt]: I see this code is unused for now. This API gonna be used lated for download tasks, right?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/GitHubReleases.kt [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    companion object {
+        fun createInstance(baseUrl: String = "https://api.github.com/"): GitHubReleasesService =
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/GitHubReleases.kt]: Why not a constructor initialization?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/run/AzureFunctionsHostConfigurationFactory.kt [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+                    true,
+                    true,
+                    RunnableProjectKind.None,
+                    "",
+                    DotNetStartBrowserParameters()
+            )
+
+    override fun createConfiguration(name: String?, template: RunConfiguration): RunConfiguration =
+            AzureFunctionsHostConfiguration(name
+                    ?: "Azure Functions host", template.project, this, createParameters(template.project))
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/run/AzureFunctionsHostConfigurationFactory.kt]: Maybe name it "Azure Functions Executable"? or "Azure Functions"? Out of scope - maybe add a separate icon with a "play" icon, like we have for .NET Executable to separate them from publish config?

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/run/AzureFunctionsHostConfigurationParameters.kt [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+                isStartJavaScriptDebugger = startBrowserParameters.withJavaScriptDebugger
+            }
+
+            BrowserStarter(runProfile, startBrowserSettings, processHandler).start()
+        }
+    }
+
+    private fun tryGetRunnableProject(): RunnableProject? {
+        val runnableProjects = project.solution.runnableProjectsModel.projects.valueOrNull
+        if (runnableProjects != null) {
```
Comment (sdubov) [inline PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/run/AzureFunctionsHostConfigurationParameters.kt]: You can use Elvis operator for such constructions: 
`... .valueOrNull ?: return null`

**Review by sdubov — APPROVED** [review]
Overall looks great! There are just several minor comments 👍

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/GitHubReleases.kt (hunk 2 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/GitHubReleases.kt]: Correct. Moving it out of this PR, makes no sense to already include it.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/GitHubReleases.kt (hunk 3 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/GitHubReleases.kt]: Also moving it out.

File: PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/run/AzureFunctionsHostConfigurationFactory.kt (hunk 4 above)
Comment (maartenba) [reply PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/run/AzureFunctionsHostConfigurationFactory.kt]: Updated, also updated icon.

**Comment by maartenba** [comment]
@sdubov Updated based on comments, merged to `191`.

**Comment by wis3guy** [comment]
@maartenba This is great stuff. When will it be released?

**Comment by maartenba** [comment]
@wis3guy Close to Rider 2019.1 :-) Adding some more functions goodies at this time.

**Pull request merged**